network/kademlia_load_balancer.go

```
@@ -0,0 +1,256 @@
+// Should return true if it must continue iterating LBBin's or stops if false.
+type LBBinConsumer func(bin LBBin) bool
+
+// KademliaLoadBalancer tries to balance request to the peers in Kademlia returning the peers sorted
+// by least recent used whenever several will be returned with the same po to a particular address.
+// The user of KademliaLoadBalancer should signal if the returned element (LBPeer) has been used with the
+// function lbPeer.AddUseCount()
+type KademliaLoadBalancer struct {
+	kademlia         KademliaBackend                    // kademlia to obtain bins of peers
+	resourceUseStats *resourceusestats.ResourceUseStats // a resourceUseStats to count uses
+	onPeerSub        *pubsubchannel.Subscription        // a pubsub channel to be notified of new peers in kademlia
+	offPeerSub       *pubsubchannel.Subscription        // a pubsub channel to be notified of removed peers in kademlia
+	quitC            chan struct{}
+
+	initCountFunc func(peer *Peer, po int) int //Function to use for initializing a new peer count
+}
+
+// Stop unsubscribe from notifiers
+func (klb *KademliaLoadBalancer) Stop() {
+	klb.onPeerSub.Unsubscribe()
+	klb.offPeerSub.Unsubscribe()
+	close(klb.quitC)
+}
+
+// EachBinNodeAddress calls EachBinDesc with the base address of kademlia (the node address)
+func (klb *KademliaLoadBalancer) EachBinNodeAddress(consumeBin LBBinConsumer) {
+	klb.EachBinDesc(klb.kademlia.BaseAddr(), consumeBin)
+}
+
+// EachBinFiltered returns all bins in descending order from the perspective of base address.
+// Only peers with the provided capabilities capKey are considered.
+// All peers in that bin will be provided to the LBBinConsumer sorted by least used first.
+func (klb *KademliaLoadBalancer) EachBinFiltered(base []byte, capKey string, consumeBin LBBinConsumer) {
+	_ = klb.kademlia.EachBinDescFiltered(base, capKey, 0, func(peerBin *PeerBin) bool {
```

Error is not handled. Even it can be only one error that is ignored now, if the function returns an error, it should be handled, at least logged.